go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll

```diff
@@ -0,0 +1,297 @@
+      )
+    }
+  }
+
+  /** Mark an empty string returned with an error as a sanitizer */
+  private class EmptyErrorSanitizer extends Sanitizer {
+    EmptyErrorSanitizer() {
+      exists(ReturnStmt r, DataFlow::CallNode c |
+        c.getTarget().hasQualifiedName("errors", "New") and
+        r.getNumChild() > 1 and
+        r.getAChild() = c.getAResult().getASuccessor*().asExpr() and
+        r.getAChild() = this.asExpr()
+      )
+    }
+  }
+
+  /** Mark any formatting string call as a sanitizer */
+  private class FormattingSanitizer extends Sanitizer {
+    FormattingSanitizer() { exists(Formatting::StringFormatCall s | s.getAResult() = this) }
+  }
+
+  private string getRandIntFunctionName() {
+    result =
+      [
+        "ExpFloat64", "Float32", "Float64", "Int", "Int31", "Int31n", "Int63", "Int63n", "Intn",
+        "NormFloat64", "Uint32", "Uint64"
+      ]
+  }
+
+  private DataFlow::CallNode getARandIntCall() {
+    result.getTarget().hasQualifiedName("math/rand", getRandIntFunctionName()) or
+    result.getTarget().(Method).hasQualifiedName("math/rand", "Rand", getRandIntFunctionName()) or
+    result.getTarget().hasQualifiedName("crypto/rand", "Int")
+  }
+
+  private DataFlow::CallNode getARandReadCall() {
+    result.getTarget().hasQualifiedName("crypto/rand", "Read")
+  }
+
+  /**
+   * Mark any taint arising from a read on a tainted slice with a random index as a
+   * sanitizer for all instances of the taint
+   */
+  private class RandSliceSanitizer extends Sanitizer {
+    RandSliceSanitizer() {
+      exists(DataFlow::Node randomValue, DataFlow::Node index |
+        // Sanitize flows like this:
+        // func GenerateCryptoString(n int) (string, error) {
+        //   const chars = "123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-"
+        //   ret := make([]byte, n)
+        //   for i := range ret {
+        //     num, err := crand.Int(crand.Reader, big.NewInt(int64(len(chars))))
+        //     if err != nil {
+        //       return "", err
+        //     }
+        //     ret[i] = chars[num.Int64()]
+        //   }
+        //   return string(ret), nil
+        // }
+        randomValue = getARandIntCall().getAResult()
+        or
+        // Sanitize flows like :
+        // func GenerateRandomString(size int) string {
+        //   var bytes = make([]byte, size)
+        //   rand.Read(bytes)
+        //   for i, x := range bytes {
+        //     bytes[i] = characters[x%byte(len(characters))]
+        //   }
+        //   return string(bytes)
+        // }
+        randomValue =
+          any(DataFlow::PostUpdateNode pun |
+            pun.getPreUpdateNode() = getARandReadCall().getArgument(0)
+          )
+      |
+        TaintTracking::localTaint(randomValue, index) and
+        (
+          this.(DataFlow::ElementReadNode).reads(_, randomValue) or
+          any(DataFlow::ElementReadNode r).reads(this, index)
+        )
```

Comment:
```suggestion
        this.(DataFlow::ElementReadNode).reads(_, index)
```

Reply:
It does not; I've re-verified that this works and will push it myself shortly along with some other last cleanups.

It's possible you're thinking of `this.(DataFlow::ElementReadNode).reads(_, randomValue)` which indeed does not work because we didn't follow the taint trail to `index`.